### ISSUE TITLE
Figure out how to swap courses

### ISSUE THREAD

**Issue by MuellMark** [MuellMark/Course-Scheduler#204]
I finished #169 more quickly than anticipated, so I can start working on this. The functionality should be there to be able to force courses to certain times. In my last hour of work I'm going to go ahead and figure out all of the requirements and start planning on how to do this.

**Comment by MuellMark** [comment]
I think the best way to do this is to have a swap courses command line param on top of new headers in the csv's. This way, I can bring the courses in, then edit the csv to have the two courses swap by forcing the opposing courses to be at different times.
I'm thinking to make it look something like this:
<img width="332" alt="Screenshot 2024-04-07 at 8 24 12 PM" src="https://github.com/MuellMark/Course-Scheduler/assets/92736704/9ffc25f2-f212-447d-8907-def7e564d2fe">
And then create something like this.
<img width="339" alt="Screenshot 2024-04-07 at 8 29 37 PM" src="https://github.com/MuellMark/Course-Scheduler/assets/92736704/a7a654b1-1147-4531-9c63-a56f65badede">

I really need to think through the flow, since I won't have access to the given course times unless a seperate files is brought in, or it runs the code once, generates the schedule to find the times, and then runs it again, which seems silly. This I think is a next week problem. For now, I can make test cases and figure that out later with colby. (ideally, the export file should be stored locally somewhere, so if we can access that it would still work)

These are the things I hope to implement first:

1. New command line parameters for for swapping a course, or changing a course to be a certain time. these will route to different methods. 
2. Forcing courses to be at certain times is already possible, I'll have to talk with colby on how to implement it
3. figure out how to alter the csv based on the export file, if this is possible.
4. Figure out what to say when a newly presented csv is infeasible.

For my last half hour, I'll get everything together for the parameters, and start thinking up how the page should look

**Comment by MuellMark** [comment]
For the site, I think under the formatted schedule, there should be two prompts: one to force a course to be a given time, and one to swap two courses. From there edits can be made one at a time. It may be tedious to do it one at a time, but I think it is very doable, and would elt the user know when things broke.

The two prompts should be something along the lines of: I would like <course drop down> to be at <times drop down>, and I would like to swap <course drop down> with <course drop down>. This should allow for sufficient edits to be made for everything.

I've added in the two new command line params, the one for force times is the same as user, but I may have to change that depeneding on what files we have our disposal, and if colby can add in the course and time to fit with my scheme. If not I'll have to make a new methods for that too.

I think we are in good shape and I laid a lot of ground work for this so that I can talk with Colby tomorrow and continue working from there

**Comment by MuellMark** [comment]
I'm going to start working on this with the assumption that I have access to the solution. If this isn;t the case we can make changes next week

**Comment by MuellMark** [comment]
I've found what may be a potential issue with the code not being able to run if there is a swapped courses section before the forced courses. I'm not sure if this is part of the courses array or the forced courses array, but it still needs to be taken care of. I need to sit down and figure out the complete flow and whether or not it will even contain this section or if it'll be removed by then.

Otherwise, I did a bit more testing and fixing csv files. Forcing courses to specific times is working, so being able to utilize it for swapping courses will work too, in theory. I'll continue working on this later

**Comment by MuellMark** [comment]
I'm going to continue working on this now

**Comment by MuellMark** [comment]
I figured out what my issues was. I was assuming that any line with a length of 2 was a forced course, which with the swapped courses is no longer true. I was going to add in extra constraints to make sure this no longer happens, but it would be much easier to just remove the lines and replaced them with forced constraints rows. So the general flow should be:

1. import both the successful schedule and the constraints
2. get the times from the successful schedule
3. store the courses to be swapped and delete them from the list
4. add in two new forced course columns coresponding to the swapped courses
5. send that to the split method

I need to look at what the website will be recieving and go from there. I may need to alter my params to cover more cases

**Comment by MuellMark** [comment]
I took a little break but I'm continuing work on this now

**Comment by MuellMark** [comment]
I went ahead and fixed the issue despite what I said above. I felt it was better to be safe than sorry.

The export file will need to have the section number, not just the course name. I went ahead and fixed that issue and started drafting up how to do the swap

This should hopefully be working soon

**Comment by MuellMark** [comment]
In order for this to work I need to rework some of the command line params, as I need different cases for how it's exporting and when there are more than one file imported (for swapping). In trying to edit this, I broke the flow and I'm trying to fix it again. 

Once this is working again I can figure out the best way to add the forced rows that I need.

**Comment by MuellMark** [comment]
The current way the params works is like this:
(<constraint file>,<file_type*>,<export_type**>,<previously exported file, for swapping>)
*  *File type can either be user (a user imported file), site (generated form the site), time (forcing a time), and swap (swapping 2 courses)
* **export either for site, or to user
I'm still working out the flow, I originally had the params being passed around, but I think making them global is much less confusing. I should be able to wrap this up shortly

**Comment by MuellMark** [comment]
I went through and fixed everything with the exporting, so now colby will be able to toggle the format in which it's exported. I may need to consider making a method that contains both, but that can wait, and should be easy.

Now that exporting is complete and everything with the command line params is sorted out, I can start fixing up everything with swapping courses

**Comment by MuellMark** [comment]
I'm storing the 2 courses and then calling the swap file to get the times for swapping. I'm still working on getting everything together. I was originally going to only get the 2nd line, but I've decided to loop thru it just in case I decided to move things around. I should get something working in the next half hour for testing.

**Comment by MuellMark** [comment]
I chose to split everything up into two while loops to avoid confusion later on in case something needs to be edited. I'm almost done with this and then I can start doing tests with this. This code really needs to be tested more thourghouly, but this can come later once a working product is complete. I'll keep working to wrap this issue up before next class so that colby and I can connect everything together

**Comment by MuellMark** [comment]
While fixing things up I found some issues with a test csv file, but it also showed me that if the code can handle issue better than I aniticpated, specifically if a course isn't specified by a faculty member. It doesn;t crash or anyhting, so I may want to add code to this, if I have time...

At this point, all of the code is in place, but it isn;t working and will need to be properly tested. This last half houor should be sufficient for this.

**Comment by MuellMark** [comment]
Everything from the code is now working, it needs to be tested more though. I haven't seen it not work but I;m not fully confident that it is completley working. That being said it is ready to be connect with colbys code tomorrow, which was the most important thing to get thru

I'm not closing this issue yet because it still needs some comments and I want it to be connected successfully to colbys code before I can called this part wrapped up

**Comment by MuellMark** [comment]
I'm meeting with Colby to connect everything together

**Comment by MuellMark** [comment]
I added a hot fix that splits an argument when there is only one param given (there are supposed to be 3-4 params but it's only registering as one). I did a quick merge to start testing on colby's side. We got it over to Colby's side but it's still not working yet. We aren't 100% sure why this is, as it could be an issue with calling the script, but it could also be something with the csvs

It was working when there were no parameters, so if it's not working I could go back and have these cases be based on the first line of the csv, assuming the csv can be editted.

**Comment by MuellMark** [comment]
We got it working but I have realized a flaw on my end. Basically, I have the file types of input and user, which makes it very confusing for colby and the users to import the csv. I'm going to go ahead and add in code to detect which type is brought in so that the method call is always the same. There will now only be 3 types: csv, time and swap, the other two we will worry about later. I'm not getting rid of them completeley because I think there is still functionality there for them.

In general, we'll need kate and yahyah to go ahead and make the schedule page nice and pretty. It is readible, but it is janky.

Once this is all nice and commented and clean we'll worry about forcing courses and swapping them. Colby and I talked about it and there a lot of ways to go about this. We could have the swapped times/ courses as command line params, or I may need to edit how the files are parsed. We should have access to the original file, so we can use it for edits, but if not, we can have the file that is passed thru with the solution containg the original script. this is a problem for a few days tho

Colby also worked on getting the schedule to look more readable before kate and yahyah come in and make it nice and pretty

Colby and I left off here, but we're in much better shape for tomorrow

**Comment by MuellMark** [comment]
I changed user and site file types into csv and then did a change for kate so that she can display the page better. We'll have to start looking into swapping, forcing and exporting later on

**Comment by MuellMark** [comment]
One thing I noticed in #240, if there is no swap tag, it breaks, I'll add that error handling real quick

**Comment by MuellMark** [comment]
Through testing, I have found that swapping is very finicky, but forced courses is working great. I'll see over the weekend if I can figure out how to fix it.

**Comment by MuellMark** [comment]
I'm going to see if I can fix what wasn't working with swapping now, before I meet with colby later today

**Comment by MuellMark** [comment]
I've been working on this for about 45 minuetes trying to add breakpoints and print statements and edits to different csv files, but I'm still at a loss. I'm going to come back to this in a bit and see if I can figure out why it's being so finicky and not working even when it's supposedly called right. I'm assumeing it broke in one of my edits for working with colby

**Comment by MuellMark** [comment]
I'm back and I'm going to continue working, Colby should arrive shortly for us to work together

**Comment by MuellMark** [comment]
For a bit more context, it seems as though the swap method is not being called at all. I wasted a lot of time trying to figure out breakpoints in vs code and going from there. I've looked at the code and nothing in the method was changed, and it was working earlier so I'm starting to look into the other methods where it's called from.

I'm still working on it, but in the mean time Colby arrived and he showed off that the csv prompts are working and are adding to the csv. We just need to have it rerun and then forcing will be working. He'll work on that while I try and fix swapping

**Comment by MuellMark** [comment]
<img width="1048" alt="Screenshot 2024-04-21 at 7 52 24 PM" src="https://github.com/MuellMark/Course-Scheduler/assets/92736704/a1322873-6887-4800-870f-a190550b739b">
Something is in fact broken with how things are added back into the csv as the forced columns are not added. 

I fixed some known issues but it still isn;t working quite right. I changed it so the forced courses are first in the list to make things easier, especially if there is not forced course restrictions. One thing I am realizing is that there could be conflicts, which I'll start taking care of.

Colby and I also did some tests of the whole flow and it is working from the csv [age exactly as I expect it. There are edits to be made, but we'll get there in a bit

**Comment by MuellMark** [comment]
The thing I noticed with colby is that I'll need to pass a few more things around (the course name and id) through the whole thing. I'll make a new issue for this later, as it;s not dire, but it is in the back of my mind

For swapping courses, they are being added to the giant list, but they aren't being consider forced courses for some reason. I have figured out the reason, lists take in any type, so when I added in the forced tag, it is a string, not a list and it's throwing off one of my conditional statements. I made that fix and now a different conditional is off. I know what to do but I'm also working with colby on the forced times section.

The forced times should almost be working, we're just looking at the flow to make sure the changes are continually saved

**Comment by MuellMark** [comment]
Forced times is in fact working now, we just needed to swap one file around so it saves the changes, which is great! There are now a bunch of things I need to do to fix things:

1. fix swap, almost done. Once my code is in place, colby can add in prompts that are almost the same as what we have to call it, Should be done by classtime ending tomorrow
2. Fix how forced columns works, because if there is a duplicate it needs to favor the newer one and get rid of the older to avoid causing problems
3. some edits to to inputs and outputs that I will specify elsewhere.

I fixed the issues with the swap method and I believe it is working. I'm going to do some clean up and test some things before I move on though.
<img width="1059" alt="Screenshot 2024-04-21 at 9 26 28 PM" src="https://github.com/MuellMark/Course-Scheduler/assets/92736704/3b8ee2cb-6226-4a9d-9c86-bcd21865b6dc">
Here we can see that cs21 and cs22 are switching. I'll do a few more tests before deleting all of the print statements

**Comment by MuellMark** [comment]
<img width="232" alt="Screenshot 2024-04-21 at 9 43 19 PM" src="https://github.com/MuellMark/Course-Scheduler/assets/92736704/515bf290-c8c7-46cb-897c-f78f11a9b5d4">

NUA and ARI are the two that were switched, which means this is working, or rather, it is not borken. I'm going to clean everything up now and make a few new issues and contiinue from there

**Comment by MuellMark** [comment]
Everything here should be complete but until it's working on the page I'll keep this issue open until everything is done. I'm going to start fixing up the the forced column issues and go from there.

**Comment by MuellMark** [comment]
I do believe I have figured out how to swap courses, as it iss working. Closing now
